src/main/java/org/example/EchoServer.java

```diff
@@ -2,18 +2,20 @@
 
 import java.io.IOException;
 import java.net.ServerSocket;
+import java.net.Socket;
 
 public class EchoServer {
 
-    public ServerSocket serverSocketCreator(int portNumber) {
-        ServerSocket serverSocket = null;
-        try {
-            serverSocket = new ServerSocket(portNumber);
-            System.out.println("--- Running echo server ---");
-        } catch (IOException ioException) {
-            ioException.printStackTrace();
-            System.exit(1);
-        }
-        return serverSocket;
+    private final ServerSocket serverSocket;
+
+    public EchoServer(ServerSocket serverSocket) {
+        this.serverSocket = serverSocket;
+    }
+
+    public void start() throws IOException {
+        System.out.println("[STARTING ECHO SERVER]");
+        Socket client = this.serverSocket.accept();
+        new EchoMessageBus(client).start();
```

Comment:
What I meant in my comment before is some code like that:
```
Socket clientSocket = this.serverSocket.accept();
ClientConnection connection = new ClientConnection(clientSocket);
String inputMessage;
while((inputMessage = connection.receiveMessage()) != null) {
                if (inputMessage.equals("end")) {
                    connection.publishMessage("[SHUTTING DOWN ECHO SERVER]");
                    connection.close();
                    break;
                }
                publishMessage(inputMessage);
}
```

Notice, that it makes visible that currently we close client socket only on graceful `end` of the session. Meaning, that on any error thrown, there is a big chance that the socket connection remains open, consuming memory and network resources of the server.


My only doubt is if it will work fine with multihreading that is in place for current implementation.
If to make it doesn't work without `.start()` which wraps `.run()` method, then we'll have to keep the loop into a wrapper, but still separating `ClientConnection` class from it
```
Socket clientSocket = this.serverSocket.accept();
ClientConnection connection = new ClientConnection(clientSocket);
ConnectionListener connectionListener = new ConnectionListener(connection);
connectionListener.start();
```

Reply:
Thanks for the explanatory comment, Sergii. I have refactored to the project to a similar format as you suggested above, let me know what you think.

Reply:
Note that some additional classes are proposed comparing to the current structure